Introduced document colors section to the font/background color plugins

msamsel opened the pull request:
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Add recently used colors section to the font color and the font background color plugins. Closes #28.

---

### Additional information

PR requires branches:
* ~PR: ckeditor/ckeditor5-ui#497 and [branch](https://github.com/ckeditor/ckeditor5-ui/tree/t/ckeditor5-font/28)~
* PR: ckeditor/ckeditor5-theme-lark#228 and [branch](https://github.com/ckeditor/ckeditor5-theme-lark/tree/t/ckeditor5-font/28)

msamsel commented:
@dkonopka I'll need to help with adding nice style for a label "recently used".
![Screen Shot 2019-04-09 at 13 57 07](https://user-images.githubusercontent.com/9017322/55798532-76d65980-5acf-11e9-8cde-f53a8e573bbd.png)

mlewand commented on src/ui/colorui.js:
```diff
@@ -117,7 +135,60 @@ export default class ColorUI extends Plugin {
+	 * @private
+	 * @param {String} color String which stores value of recently applied color
+	 */
+	addColorToDocumentColors( color ) {
```
I tend to see across the entire PR that your code spacing style here is a little clunky. You're not putting sufficient spacing in your new code.

Basically you should give your code a little more "breathing room".

This function is an example, it would be nice to put a single line of spacing between const initialization to separate the code initialization and the actual meat.

Think of it as about a book. It would be difficult to read a book with no spacing at all. Instead it is a common practice to put one thought in a single paragraph. Then make a spacing and start new paragraph/thought.

Similar thing happens with lyrics.

It's a subjective thing, yes, but you can see a pretty consistent patterns across our codebase - so our devs tend to "feel it".

mlewand replied:
Although I see you added some spacing in the API docs - that's cool. 👍

mlewand commented on tests/manual/font-color.html:
```diff
@@ -10,3 +10,15 @@
 	<p><span style="color:hsl(270, 75%, 60%);background-color:#d82;">10. color:hsl(270, 75%, 60%);background-color:#d82;</span></p>
 </div>
+
+<div id="color-box" style="margin-top: 10px;border:1px solid black;padding:5px;">
```
This inline style formatting is inconsistent with the rest of codebase. Actually only font plugin doesn't put a space between `;` and a next property so it might require also a more borad change in a separate PR. See:

* `engine/tests/view/manual/focusobserver.html`
* `markdown-gfm/tests/gfmdataprocessor/manual/gfmdataprocessor.html`
* `media-embed/tests/manual/mediaembed.html`
* `typing/tests/manual/rtl.html`
* `utils/tests/manual/position/position.html`

Also last `;` might be removed, but that's just a detail.

mlewand commented on src/fontbackgroundcolor.js:
```diff
@@ -140,3 +140,11 @@ export default class FontBackgroundColor extends Plugin {
+/**
+ * Determines maximum number of available document colors. If not defined,
+ * then it equals to {@link module:font/fontbackgroundcolor~FontBackgroundColorConfig#columns} value.
+ * If set to `0`, then document colors feature will be disabled.
```
```suggestion
 * Setting it to `0` will disable document colors feature.
```

mlewand replied:
Change should also be applied to the occurrence in the font color config.

mlewand replied:
It would be nice to add a listing with this property being used. Also it would be nice to put information about the default value as the last information (but before the example snippet).

mlewand commented on src/ui/colorui.js:
```diff
@@ -117,7 +135,60 @@ export default class ColorUI extends Plugin {
+					if ( val ) {
+						const model = editor.model;
+						const document = model.document;
+						const rootNames = document.getRootNames();
```
This variable should be inlined, it's used only once in for of loop. While extracting `maxCount` (that is in the following line and also occurs only once) sounds like a good idea (increases readability) this variable doesn't have this benefit.

mlewand commented on src/ui/colorui.js:
```diff
@@ -117,7 +135,60 @@ export default class ColorUI extends Plugin {
+									}
+								}
+							}
+							if ( documentColors.length >= maxCount ) {
```
I don't see any reason for this check here. New colors are only added in line https://github.com/ckeditor/ckeditor5-font/blob/47da79a/src/ui/colorui.js#L159 and current `documentColors` size is correctly checked afterwards.

This check is redundant.

msamsel replied:
This check was made to exit from outer `for` loop, to not iterate over other roots in editor, when already all document colors are found.
I'll extract this part to function and made single `return` statement to increase its readability and remove additional `if` statement.

mlewand replied:
Sounds good 👍

msamsel commented:
> Three unit tests are failing for me

You need to switch branch of `ckeditor5-ui` as I described in "Additional information" in PR. Should I make additional PR for those 2 additional branches and link them somehow here?

mlewand commented on src/fontbackgroundcolor.js:
````diff
@@ -132,11 +132,51 @@ export default class FontBackgroundColor extends Plugin {
+ * By default it equals to {@link module:font/fontbackgroundcolor~FontBackgroundColorConfig#columns} value.
+ *
+ * Examples:
+ * ```js
````
We currently are not using fenced code in our JSDocs. So let's keep code aligned with the rest of the project.

mlewand commented on src/ui/colorui.js:
```diff
@@ -117,7 +135,72 @@ export default class ColorUI extends Plugin {
+	 *
+	 * @private
+	 */
+	findAndSetDocumentColors() {
```
Private methods should be prefixed with an underscore.

mlewand replied:
Better name will be `updateDocumentColors()`.

mlewand commented on src/ui/colorui.js:
```diff
@@ -117,7 +135,72 @@ export default class ColorUI extends Plugin {
+	 * Method scans through editor's content and search for text node attributes with name defined in {@link #commandName}.
+	 * Found entries are set as document colors.
+	 *
+	 * All previously stored document colors will be lost with that process.
```
```suggestion
	 * All the previously stored document colors will be lost in the process.
```

mlewand commented on src/ui/colorui.js:
```diff
@@ -117,7 +135,72 @@ export default class ColorUI extends Plugin {
+	 *
+	 * All previously stored document colors will be lost with that process.
+	 *
+	 * Method is used to determines document colors when UI dropdown is opened.
```
This line is not needed. It is implementation detail where a given method is used. It adds more coupling between code and documentation, so you would need to remember to update one when changing the other.

mlewand commented on src/ui/colorui.js:
```diff
@@ -117,7 +135,72 @@ export default class ColorUI extends Plugin {
 	}
+
+	/**
+	 * Method scans through editor's content and search for text node attributes with name defined in {@link #commandName}.
```
```suggestion
	 * Method scans through editor's content and searches for text node attributes with the name defined in {@link #commandName}.
```

mlewand commented on src/ui/colorui.js:
```diff
@@ -117,7 +135,72 @@ export default class ColorUI extends Plugin {
+	}
+
+	/**
+	 * Method tries to find a color on predefined color list to prevent of creating custom color from it.
```
It's enough to say:

```suggestion
	 * Method adds the `color` to document colors list. If possible, it will attempt to use data from the {@link #colorDefinitions} (label, color options).
```

The rest of header description should be removed.

Your proposal included too much details and a little bit of repetition.

mlewand commented on src/utils.js:
```diff
@@ -109,11 +109,13 @@ export function normalizeColorOptions( options ) {
  * @param {Array.<module:ui/colorgrid/colorgrid~ColorDefinition>} config.colors An array with definitions
  * representing colors to be displayed in the color table.
  * @param {String} config.removeButtonLabel The label for the button responsible for removing the color.
+ * @param {String} config.documentColorsLabel The label for the section with document colors.
```
```suggestion
 * @param {String} config.documentColorsLabel The label of a section with document colors.
```

mlewand commented on src/ui/colortableview.js:
```diff
@@ -122,7 +151,11 @@ export default class ColorTableView extends View {
 
 		this.items.add( this.removeColorButton() );
-		this.items.add( this.createStaticColorTable() );
+		this.items.add( this.createStaticColorGrid() );
```
I don't see a reason to change the name of this function.

msamsel replied:
It is more meaningful for me, as this method create instance of [ColorGridView](https://ckeditor.com/docs/ckeditor5/latest/api/module_ui_colorgrid_colorgrid-ColorGridView.html)
That's why I've made this change in name. Using `grid` word is coherent with naming in returned type.

This same applies to `documentColors` section.

mlewand commented on src/ui/colortableview.js:
```diff
@@ -165,6 +198,63 @@ export default class ColorTableView extends View {
+	 *
+	 * @private
+	 */
+	createDocumentColors() {
```
Should be called `createDocumentColorsTable()` for consistency.

mlewand replied:
Missing return type.

msamsel replied:
I prefer to name it `_createDocumentColorsGrid` to be coherent with static colors.
Reason is the same as described in: https://github.com/ckeditor/ckeditor5-font/pull/39#discussion_r283345747

mlewand commented on src/ui/colortableview.js:
```diff
@@ -165,6 +198,63 @@ export default class ColorTableView extends View {
 	}
 
+	/**
+	 * Adds document colors section view and binds it to {@link #documentColors}.
```
It _creates_ it it does not _add_ it.

mlewand commented on src/ui/colortableview.js:
```diff
@@ -165,6 +198,63 @@ export default class ColorTableView extends View {
+					} );
+				}
+
+				if ( colorObj.options && colorObj.options.isEnabled === false ) {
```
I don't understand what does this line do. I don't see `options.isEnabled` documented.

msamsel replied:
Good catch. It seems to remain when I removed 'recently used colors' in favour of a document colors. There was displayed inactive buttons used as colors indicators.

mlewand commented on src/ui/colorui.js:
```diff
@@ -18,6 +18,10 @@ import {
  *
+ * In case that `documentColors` for given plugins are not defined in the configuration,
+ * then by default its value equals to its `column` value. This results with
+ * displaying one row of document colors.
```
This paragraph is irrelevant to the `ColorUI` class. It should be (and is) documented in the config property.

msamsel commented:
@mlewand I've made fixes for mentioned parts except situation described here: https://github.com/ckeditor/ckeditor5-font/pull/39#discussion_r283345747

If you feel that still those methods should be renamed to `table` please give me an info.

mlewand commented:
I checked the tests and label with proper branch in the `theme-lark` and it does fix the tests / missing label. Sorry for the confusion.

msamsel commented:
@mlewand I linked PRs to 2 other repositories in PR description at the top.

mlewand commented on src/ui/colortableview.js:
```diff
@@ -152,45 +215,69 @@ export default class ColorTableView extends View {
+		);
+
+		this.documentColors.on( 'add', ( evt, item ) => {
+			const duplicates = this.documentColors.filter( element => element.color === item.color, this );
```
This logic is strictly related to the document colors collection. Given the fact that it make sense to implement a `isEmpty` observable for this particular collection (to simplify the binding) it is a good time to add a specialized class for that.

mlewand commented on src/ui/colortableview.js:
```diff
@@ -152,45 +215,69 @@ export default class ColorTableView extends View {
+					this.fire( 'execute', {
+						value: colorObj.color,
+						label: colorObj.label,
+						options: Object.assign( {}, colorObj.options )
```
Note that in my proposed changes there's no longer a need for `colorDefinition.options` so it's likely this code bit should be changed.

mlewand commented on src/ui/colortableview.js:
```diff
@@ -152,45 +215,69 @@ export default class ColorTableView extends View {
-	render() {
-		super.render();
+	_createDocumentColorsGrid() {
+		const documentColors = new ColorGridView( this.locale, {
```
`ColorGridView` class should not maintain the label. The label should be simply added here in `ColorTableView` as a child before the document colors grid. See https://github.com/ckeditor/ckeditor5-ui/pull/497#issuecomment-493933922

So the label property will no longer be needed

mlewand commented on src/ui/colortableview.js:
```diff
@@ -83,13 +88,37 @@ export default class ColorTableView extends View {
+		 *
+		 * @type {String}
+		 */
+		this.documentColorsLabel = documentColorsLabel;
```
There's no need to expose this property. Document colors label will be used in the constructor and it is (for now) no longer needed afterwards.

Let's not expose it.

mlewand commented on src/ui/colorui.js:
```diff
@@ -59,15 +61,24 @@ export default class ColorUI extends Plugin {
+		 *
+		 * @type {module:font/ui/colortableview~ColorTableView}
+		 */
+		this.colorTableView;
```
We're not declarations [like that](https://github.com/ckeditor/ckeditor5-font/blob/5f512d754067e61ed5c7b9d3e30030e3c739dfcd/src/ui/colorui.js#L81). Simply use @member api tag to declare a member. Like here: https://github.com/ckeditor/ckeditor5-font/blob/5f512d754067e61ed5c7b9d3e30030e3c739dfcd/src/fontcommand.js#L32

msamsel replied:
I don't mind to change it.

However there is plenty of examples where we declare class properties in such way:
https://github.com/ckeditor/ckeditor5-engine/blob/master/src/model/document.js#L44-L96
or:
https://github.com/ckeditor/ckeditor5-font/blob/d4570220c7d7a0e1e8b5817f81c65bdf8be1a98e/src/ui/colorui.js#L75

`@type` seems to be nicer here, as you don't have possibility to make any typo mistake and you don't need to remember to update docs if parameter name will change.

mlewand commented on src/ui/colorui.js:
```diff
@@ -117,7 +131,101 @@ export default class ColorUI extends Plugin {
+	 */
+	_updateSelectedColors() {
+		const staticColorsGrid = this.colorTableView.items.get( 1 );
+		const documentColorGrid = this.colorTableView.items.get( 2 );
```
This (and above) check is insufficient, you should retrieve such information from your data/model rather than view. View is **very** likely to change.

Try to avoid relying on view structure if possible.

mlewand commented on src/ui/colorui.js:
```diff
@@ -117,7 +131,101 @@ export default class ColorUI extends Plugin {
+		const selectedColor = this.colorTableView.selectedColor;
+
+		if ( documentColorGrid ) {
+			if ( isInDocumentColors( documentColorGrid.items, selectedColor ) ) {
```
This code is most certainly in domain of ColorTableView, rather than ColorUI.

Just look how many `staticColorsGrid` or `documentColorGrid` references you're using here. That's already a sniff that something should be implemented a different way.

msamsel commented:
@mlewand I cherry pick your commits to `t/28` and apply changes requested in Review.
Please notice that change from https://github.com/ckeditor/ckeditor5-theme-lark/pull/228 is still required for good looking dropdown.

mlewand commented on src/documentcolorscollection.js:
```diff
@@ -0,0 +1,77 @@
+import mix from '@ckeditor/ckeditor5-utils/src/mix';
+
+/**
+ * @module font/documentcolorscollection
```
I have just noticed that we'll have to adjust the name to singular form, so that it follows rest of the code (e.g. PluginCollection, ViewCollection, MarkerCollection).

So it should become `DocumentColorCollection`.

mlewand commented on src/ui/colortableview.js:
```diff
@@ -121,16 +163,122 @@ export default class ColorTableView extends View {
+		this.items.add( this.staticColorsGrid );
+
+		if ( documentColorsCount ) {
+			// Create Label for Document Colors
```
Sentence in a comment should be finished with a dot.

mlewand replied:
```suggestion
			// Create a label for document colors.
```

mlewand commented on src/ui/colortableview.js:
```diff
@@ -121,16 +163,122 @@ export default class ColorTableView extends View {
+
+			this.items.add( label );
+
+			// Create Document Colors
```
This comment line is actually stray, should be removed.

mlewand commented on src/ui/colortableview.js:
```diff
@@ -121,16 +163,122 @@ export default class ColorTableView extends View {
+	 * All the previously stored document colors will be lost in the process.
+	 *
+	 * @param {module:engine/model/model~Model} model used as source to obtain document colors
+	 * @param {String} componentName determines what is name of related model's attribute for given dropdown
```
`componentName` - this property is actually an attribute name (which value is added to the document colors section).

`componentName` is a misleading name, and unrelated to the logic. It should be `colorAttributeName` or `attributeName`.

mlewand commented on src/ui/colorui.js:
```diff
@@ -117,6 +131,15 @@ export default class ColorUI extends Plugin {
 				editor.editing.view.focus();
 			} );
 
+			dropdownView.on( 'change:isOpen', ( evt, name, val ) => {
```
We decided not to use shorten variable values in v5. So if anything should be `value`. But even nicer var name would be `isVisible`. That would clearly indicate what it means.

mlewand commented on tests/manual/font-color.js:
```diff
@@ -38,3 +38,15 @@ ClassicEditor
+	};
+}
+
+document.querySelector( '#color-box #color' ).addEventListener( 'change', updateText( 'color' ) );
```
You can use ancient (yet shorter and designed specifically for that) `document.getElementById( 'color' )`. 🙂

mlewand commented on tests/manual/font-color.html:
```diff
@@ -10,3 +10,15 @@
+<div id="color-box" style="margin-top: 10px; border: 1px solid black; padding: 5px">
+	<label>
+		Color:
+		<input type="text" name="color" id="color">
```
It is not clear how to use it, someone who sees it for the first time has no clue how to use it.

mlewand replied:
Adding a placeholder text and maybe some title to the component would be enough.

mlewand commented on src/documentcolorscollection.js:
```diff
@@ -0,0 +1,77 @@
+	 * @param {String} color
+	 */
+	hasColor( color ) {
+		return !!this.find( item => item.color === color );
```
Missing return type.

mlewand commented on docs/_snippets/features/font.html:
```diff
@@ -23,4 +23,7 @@
 	</p>
 
+	<p>
+		<span style="color:#505050;background-color:rgb(75, 255, 190)">This text has custom color used for text and background. Those colors are available in 'recently used' section.</span>
```
There's no "recently used" section.

mlewand commented:
Pushed couple of fixes from my local files.

Reinmar commented on docs/features/font.md:
```diff
@@ -252,6 +252,33 @@ ClassicEditor
 
+### Documents colors
+
+In addition to predefined list of colors, the plugin also includes document colors feature. It lists the colors used in the document and lists them in the order of occurrence. Duplicated colors are not inserted multiple times. By default number of document colors is limited, but you can adjust it (or disable it) using {@link module:font/fontcolor~FontColorConfig#documentColors `fontColor.document ...
```
> In addition to predefined list of colors, the plugin also includes document colors feature.

@AnnaTomanek could you check this part? It doesn't read well.

Reinmar replied:
For me the plugin implements something and the UI includes something.

msamsel replied:
FYI: I rewrite this part a little bit.

Reinmar commented on docs/features/font.md:
```diff
@@ -252,6 +252,33 @@ ClassicEditor
+ClassicEditor
+	.create( document.querySelector( '#editor' ), {
+		fontColor: {
+			documentColors: 10,
```
I don't like a context-less sample like this. Why would anyone change the document colors to 10? Either they changed the number of columns in general (in which it'd be good to show and explain this particular case) and need to change the number of document colors to 10 too (in fact, that should happen automatically). Or they want to have e.g. two rows. Either way, showing both options here would be preferred.

Reinmar replied:
@mlewand checked that `documentColors` default to `columns` so you only need to change this value if you want to have multiple rows. Let's explain exactly that and mention `columns` as well.

Reinmar replied:
BTW, I can see in the snippet above (`custom-font-color-and-background-color-options.js`) that we modify both values. Why isn't this reflected anywhere else in this document?

msamsel replied:
I add more examples how to configure document colors section and columns together in "Changing the geometry of the color grid" section.

msamsel commented:
@Reinmar, I updated bunch of those docs. Information about correlation between `columns` and `documentColors` is now moved completely to "Changing the geometry of the color grid" section. I hope there will be more logic in description of document colors plugin.
